R/finalize_plot.R

```diff
@@ -434,3 +463,75 @@ finalize_plot <- function(input_plot = NULL,
     return(final_plot)
   }
 }
+
+
+
+#' @noRd
+# Function to create plot object with left aligned legend on top
+buildChart <- function(input_plot,
+                       plot_constants,
+                       legend_build) {
+
+  # in safe mode, don't extract legend
+  if(legend_build == "safe"){
+    return(ggplotGrob(input_plot))
+  }
+
+  # in no legend mode, remove legend altogether
+  if(legend_build == "none"){
+    output_plot <- input_plot + theme(legend.position = "none")
+    return(ggplotGrob(output_plot))
+  }
+
+  # Reformat plot
+  ## IS ANY OF THIS ACTUALLY NECESSARY? CAN WE MOVE MORE OF THIS OVER TO THEME_CMAP?
+  format_plot <- input_plot + theme(
+    # ensure legend is left aligned
+    legend.position = "left",
+    # update margins to account for possible indent
+    plot.margin = grid::unit(plot_constants$padding_plot,"bigpts"),
+    legend.margin = margin(t = plot_constants$padding_legend[1],
+                           r = plot_constants$padding_legend[2],
+                           b = plot_constants$padding_legend[3],
+                           l = plot_constants$padding_legend[4] + plot_constants$legend_indent,
+                           "bigpts")
+    )
+
+  # Extract the legend
+  legend <- ggpubr::get_legend(format_plot)
+
+  # extract the total number of "heights" in the legend (5 is standard for one
+  # legend, with each additional legend adding two additional height elements to
+  # the total)
+
+  legend_total <- length(legend$heights)
+
+  # extract height of legend object within ggplot plot. For plots with only one
+  # legend, this returns the 3rd element, which is the height of the legend
+  # component. For plots with two or more, it returns the sum of the heights of
+  # every element from the third element to the third-to-last element, which
+  # includes both text/key heights and buffers between legends)
+  legend_height <- grid::convertUnit(sum(legend$heights[3:(legend_total - 2)]),
+                                     "bigpts",
+                                     valueOnly = TRUE)
+
+  plot_height <- plot_constants$plotbox_height - legend_height - plot_constants$margin_v5
+
+  # Assemble a combined grob
+  built <- gridExtra::arrangeGrob(
+    # the legend, as a grob
+    grob(legend)[[1]],
```

Comment:
I don't think it needs it. I've removed as of commit 7cec129

Reply:
Ok, if it works without it, great. I remember having trouble getting the legend to appear and when I changed this code it worked, but that could have been the result of some other changes as well.

Reply:
<<shrug>>. Lets see if any problems show up in further testing--reverting to this might be needed. But, as I understand it, a tablegrob is a grob and should be able to be plotted without additional conversion.